zerver/actions/users.py

```diff
@@ -180,89 +112,116 @@ def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
-        ).update(user_profile=temp_replacement_user)
-        user_profile.delete()
-
-        replacement_user = create_user(
-            force_id=user_id,
             email=Address(
                 username=f"deleteduser{user_id}", domain=get_fake_email_domain(realm.host)
             ).addr_spec,
             password=None,
-            realm=realm,
-            full_name=f"Deleted User {user_id}",
             active=False,
+            bot_type=user_profile.bot_type,
+            full_name=f"Deleted User {user_id}",
+            bot_owner=user_profile.bot_owner,
             is_mirror_dummy=True,
-            force_date_joined=date_joined,
-            create_personal_recipient=False,
+            tos_version=user_profile.tos_version,
+            # These required arguments should get default values configured on the UserProfile model,
+            # to overwrite the values of user_profile.
+            timezone=UserProfile._meta.get_field("timezone").get_default(),
+            default_language=UserProfile._meta.get_field("default_language").get_default(),
+            email_address_visibility=UserBaseSettings.EMAIL_ADDRESS_VISIBILITY_EVERYONE,
         )
-        # We don't delete the personal recipient to preserve  personal messages!
-        # Now, the personal recipient belong to replacement_user, because
-        # personal_recipient.type_id is equal to replacement_user.id.
-        replacement_user.recipient = personal_recipient
-        replacement_user.save(update_fields=["recipient"])
-
-        # Uses index: zerver_message_realm_sender_recipient (prefix)
-        Message.objects.filter(realm_id=realm.id, sender=temp_replacement_user).update(
-            sender=replacement_user
+
+        overwrite_data = model_to_dict(
+            temp_replacement_user,
+            exclude=[
+                "id",
+                "date_joined",
+                "realm",
+                "recipient",
+                "uuid",
+                "last_login",
+                # These come from Django and aren't used by Zulip. We need to exclude them as they are m2m fields
+                # and as such, would break the code below.
+                "groups",
+                "user_permissions",
+            ],
         )
+
+        UserProfile.objects.filter(id=user_profile.id).update(**overwrite_data)
+        user_profile.refresh_from_db()
+
+        update_all_subscriber_counts_for_user(user_profile=user_profile, direction=-1)
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.STREAM
+        ).delete()
+
         Subscription.objects.filter(
-            user_profile=temp_replacement_user, recipient__type=Recipient.DIRECT_MESSAGE_GROUP
-        ).update(user_profile=replacement_user, is_user_active=replacement_user.is_active)
-        temp_replacement_user.delete()
+            user_profile=user_profile, recipient__type=Recipient.DIRECT_MESSAGE_GROUP
+        ).update(is_user_active=False)
```

Comment:
Question: Does other code maintain `is_user_active` at all for `Subscription` objects to `DIRECT_MESSAGE_GROUP` objects?